cmake refactoring + hunterizations

headupinclouds opened the pull request:
* add huntergate
* use find_package instead of add_subdirectory fo
   + rabit
   + dmlc-core
   + cub
   + nccl (commented)
* replace global cmake functions with target_ calls to accomodate the migration to pkgconfig targets
* make some options specific to the non R library build
* add XGBOOST_BUILD_SHARED
* add XGBOOST_BUILD_EXAMPLE
* add XGBOOST_BUILD_TEST
* add pkgconfig installation
* generate empty *.cc files required for add_library calls (Xcode generator)

headupinclouds committed:
cmake refactoring + hunterizations

* add huntergate
* use find_package instead of add_subdirectory fo
   + rabit
   + dmlc-core
   + cub
   + nccl (commented)
* replace global cmake functions with target_ calls to accomodate the migration to pkgconfig targets
* make some options specific to the non R library build
* add XGBOOST_BUILD_SHARED
* add XGBOOST_BUILD_EXAMPLE
* add XGBOOST_BUILD_TEST
* add pkgconfig installation
* generate empty *.cc files required for add_library calls (Xcode generator)

commit 6b659e2f7ea39b701d4dea02af2b564e724d8c45
Author: David Hirvonen <dhirvonen@elucideye.com>
Date:   Sat Jan 6 12:31:40 2018 -0500

    cmake fixes

    * don’t force release build
    * add temporary empty cc file for static lib creation
    * remove set_output_directory

commit 014a754bf961621b3d0762c164abe430d28ea810
Author: David Hirvonen <dhirvonen@elucideye.com>
Date:   Sat Jan 6 12:30:31 2018 -0500

    reset submodule

commit 0cd092c3b9c9fbdc92d0d42b5eff01893504f5e9
Author: David Hirvonen <dhirvonen@elucideye.com>
Date:   Sat Jan 6 12:21:48 2018 -0500

    debug submodule

commit 5920dfee9758b2f622828c7dfed19f50b9863e49
Author: David Hirvonen <dhirvonen@elucideye.com>
Date:   Sat Jan 6 02:40:08 2018 -0500

    hunterize wip

headupinclouds committed:
newline

headupinclouds commented on CMakeLists.txt:
```diff
@@ -1,21 +1,30 @@
-set(GPU_COMPUTE_VER 35;50;52;60;61 CACHE STRING
-  "Space separated list of compute versions to be built against")
+
+# TODO
```
@ruslo: FYI, I've tested linking `nccl` in xgboost builds locally with Hunter via `GIT_SUBMODULE` using:

```
xgboost_opts=(
    GOOGLE_TEST=ON
    USE_CUDA=ON
    CUDA_TOOLKIT_ROOT_DIR=/Developer/NVIDIA/CUDA-9.0
    XGBOOST_BUILD_SHARED=OFF
    R_LIB=OFF
    JVM_BINDINGS=OFF
)

polly.py --toolchain xcode --verbose --reconfig --config Debug --fwd ${xgboost_opts[@]} ${*}
```

Using CUDA's proprietary SYSTEM libs introduces a wrinkle.  I believe this typically needs some custom arguments like the `--fwd CUDA_TOOLKIT_ROOT_DIR=/Developer/NVIDIA/CUDA-9.0` to work on most platforms.  This doesn't currently fit with the hunter `./jenkins.py` test.  I don't have a multi-gpu rig for testing this anyway, so I've left nccl stuff stubbed out in xgboost for the initial release until the CUDA stuff can be resolved.

ruslo replied:
> Using CUDA's proprietary SYSTEM libs introduces a wrinkle. I believe this typically needs some custom arguments like the --fwd CUDA_TOOLKIT_ROOT_DIR=/Developer/NVIDIA/CUDA-9.0 to work on most platforms

I haven't used it so I don't know what peculiarities it has. If we can download and build stuff without buying/registering license then it's not a problem for Hunter, we can add binary package. It will be user's responsibility to deal with runtime. The question here will be the size of download though.

The second option is to make a wrapper package that can find and import CUDA libraries. Will needed to set default version in `configs/default.cmake` and save version of CUDA in Config-ID to allow multiple non-conflict configurations.

headupinclouds replied:
I'm going to "promote" the  CUDA discussion to the main hunter issues.  AFAIK, CUDA itself can be managed by hunter using "plain" downloads.

For example:

https://developer.nvidia.com/cuda-downloads?target_os=MacOSX&target_arch=x86_64&target_version=1012&target_type=dmgnetwork

There are other NVIDIA libs, such as `nccl v2.0` which require registration.

Managing these installations with hunter seems to make sense.

headupinclouds replied:
Initial (build) tests will have to be CPU only.

headupinclouds commented on CMakeLists.txt:
```diff
@@ -84,69 +88,92 @@ file(GLOB_RECURSE CUDA_SOURCES
-
-  cuda_add_library(gpuxgboost ${CUDA_SOURCES} STATIC)
+  # Note: Some CMake generators won't support libraries without at least 1 source file:
+  set(xgboost_gpu_tmp ${CMAKE_CURRENT_BINARY_DIR}/tmp/xgboost_gpu.cc)
```
@ruslo: I'm starting to use this as a common CMake pattern for `add_library` and `cuda_add_library` calls.  I believe it handles dependencies correctly (`copy_if_different`) and per documentation if the `OUTPUT` file is explicitly included in a target's sources it should all work out fine.  Let me know if you see any improvements, since it happens 2x here.  The Xcode generator (at least) requires at least one c++ file for both `cuds_add_library` calls (to complement a list of `*.cu` files) and for `OBJECT` libraries.

ruslo replied:
Is is for Xcode limitation of at least one real file in sources? Why not just use some dummy file?

headupinclouds replied:
It seemed to be less intrusive address this directly in CMakeLists.txt with something like `file(WRITE "" null.cc)`, but when the extra `add_custom_command` logic is added to prevent rebuilds, it probably is cleaner just to add empty files as needed.

headupinclouds commented on CMakeLists.txt:
```diff
@@ -40,36 +49,31 @@ else()
+  list(APPEND xgboost_defs
+    XGBOOST_STRICT_R_MODE=1
+    XGBOOST_CUSTOMIZE_GLOBAL_PRNG=1
+    DMLC_LOG_BEFORE_THROW=0  # DMLC: should be passed in config
```
These flags seem to be preferred for the R package.  I wouldn't use this build mode, but it is quite popular in the data science community (many xgboost users).  Prior to hunter + find_package() configurations the `rabit` and `dmlc-core` libs were built using `add_subdirectory()`, so these flags were globally defined and would be inherited by those libs.  In the package management case, I suppose we would have to introduce CMake options in `dmlc-core` and `rabit` to allow setting these from a top level `config.cmake` as needed.  I did test the `R` build.

ruslo replied:
> I suppose we would have to introduce CMake options in dmlc-core and rabit to allow setting these from a top level config.cmake as needed

Yep

ruslo commented on cmake/Config.cmake.in:
```diff
@@ -0,0 +1,26 @@
+endif()
+
+if(@JVM_BINDINGS@)
+  find_package(JNI QUIET REQUIRED)
```
I think `QUIET` and `REQUIRED` are not friends :)

headupinclouds replied:
Copy and paste from upstream.  That does seem like a conflict :smile:

ruslo commented on cmake/Config.cmake.in:
```diff
@@ -0,0 +1,26 @@
+endif()
+
+if(@R_LIB@)
+  find_package(LibR REQUIRED)
```
Will produce an error on user's side since custom `FindLibR.cmake` used internally.

headupinclouds replied:
Good catch.  I had assumed `find_package(LibR)` was supported by CMake natively.

ruslo commented on cmake/Config.cmake.in:
```diff
@@ -0,0 +1,26 @@
+check_required_components("@PROJECT_NAME@")
+
+if(@OpenMP_CXX_FOUND@ OR @OPENMP_FOUND@)
+  find_package(OpenMP)
```
`REQUIRED` missing? If it's pure compiler feature then it's okay. If it depends on libraries installed on system then it should be controlled with option `XGBOOST_USE_OPENMP`.

headupinclouds replied:
> REQUIRED missing? If it's pure compiler feature then it's okay.

I believe it is always a compiler feature.  The upstream CMakeLists.txt runs `find_package(OpenMP)` and uses that as a test for availability.

ruslo commented on CMakeLists.txt:
```diff
@@ -157,94 +184,204 @@ if(R_LIB)
-  add_dependencies(xgboost runxgboost)
+  # JVM
+  if(JVM_BINDINGS)
+    find_package(JNI QUIET REQUIRED)
```
I think `QUIET` can be removed.

headupinclouds replied:
Will do.

ruslo commented on CMakeLists.txt:
```diff
@@ -157,94 +184,204 @@ if(R_LIB)
 
-  # Shared library
-  add_library(xgboost SHARED $<TARGET_OBJECTS:objxgboost>)
+  option(XGBOOST_BUILD_SHARED "Build shared library" ON)
```
It's better to use standard `BUILD_SHARED_LIBS`.

headupinclouds replied:
Ok.  We can specific `STATIC` specifically for the utility libraries, such as `gpuxgboost`, and control the `xgboost` target using CMake's `BUILD_SHARED_LIBS`.

headupinclouds committed:
replace generated null files with real files

headupinclouds committed:
remove QUIET for find_package(JNI)

headupinclouds committed:
remove XGBOOST_BUILD_SHARED in favor of CMake’s default BUILD_SHARED_LIBS

Note: In XCode testxgboost was failing with `BUILD_SHARED_LIBS=ON` due to

```
Check failed: fmap_.count(name) == 0U (1 vs. 0) grow_gpu_hist already registered
```

We introduce a preprocessor definition `XGBOOST_SKIP_REGISTER` to optionally skip the registration in the case that the cuda file is included

#include "../../../src/tree/updater_gpu_hist.cu"

headupinclouds commented:
I believe all points are addressed with the exception of the LibR issue mentioned [here](https://github.com/hunter-packages/xgboost/pull/10#discussion_r160039565).  I don't see any such support in CMake. 

I see 3 options:
* install FindLibR.cmake ->  <prefix>/lib/cmake/xgboost/ (???)
* add R to hunter https://cran.cnr.berkeley.edu
* disable R support for the hunter xgboost package

Shall we install the `FindLibR.cmake` file?

ruslo commented:
> Shall we install the FindLibR.cmake file?

No. 1) `FindLibR.cmake` is not designed to be installed in general 2) We will have conflict if other package will want to install this file too.

> add R to hunter https://cran.cnr.berkeley.edu

It's the best option as usual.

> disable R support for the hunter xgboost package

Yes, easy workaround.

Another option is to add `FindLibR.cmake` to Hunter:
* https://github.com/ruslo/hunter/tree/master/cmake/find

Quick dirty hack.

headupinclouds commented:
> Another option is to add FindLibR.cmake to Hunter

Okay, I'll send a PR with this module for now (perhaps until R is added), and will make sure R is `OFF` by default in xgboost.  I'm also fine disabling R support, but I think this would be a nice option for xgboost users.

[UPDATE] https://github.com/ruslo/hunter/pull/1294

I'll merge this xgboost PR and get the package prepared in parallel with the above PR.

headupinclouds commented:
Thanks for the review/fixes!
